### PR TITLE
Allow multiple content elements in a single container

### PR DESCRIPTION
**Pull request by b-pass**
## Description

This change adds a new IListProvider, called CMultiProvider, which aggregates several independent &lt;content&gt; blocks (either DirectoryProviders or StaticProviders or both).
## Motivation and Context

DirectoryProviders (dynamically filled &lt;content&gt; blocks) are a powerful feature.  But currently you can only fill a list with a single directory at a time, and you also cannot add other static elements to a dynamic list.  For example, you might like to add some navigational list entries at the top of a list, and have it fill with dynamic content below those static entries.  Or maybe you would like to combine two different directories in one list for some reason.  This was hinted as a potential new feature in 2014, but it seems it never got added: http://forum.kodi.tv/showthread.php?tid=176864
## How Has This Been Tested?

I tested this by trying a few different combinations of multiple &lt;content&gt; blocks.
## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
## Checklist:

- [X] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by HitcherUK** [comment]
Nice, thanks.

**Comment by BigNoid** [comment]
Nice, tested and works great. Thx for this addition!
You can pick https://github.com/BigNoid/xbmc/commit/241de15944ef145150a3a07f190cd5d731d5def4 for Xcode sync

**Comment by b-pass** [comment]
Picked and Squashed, thanks!

File: xbmc/listproviders/MultiProvider.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+public:
+  CMultiProvider(const TiXmlNode *first, int parentID);
+  
+  virtual bool Update(bool forceRefresh);
+  virtual void Fetch(std::vector<CGUIListItemPtr> &items) const;
+  virtual bool IsUpdating() const;
+  virtual void Reset(bool immediately = false);
+  virtual bool OnClick(const CGUIListItemPtr &item);
+  virtual bool OnInfo(const CGUIListItemPtr &item);
+  virtual bool OnContextMenu(const CGUIListItemPtr &item);
```
Comment (garbear) [inline xbmc/listproviders/MultiProvider.h]: can you add the override keyword on these functions?

**Comment by xhaggi** [comment]
looks good +1

**Comment by da-anda** [comment]
this would be great for some home screen widgets on Estuary in v17. Are there any major concerns to hold it back for v17 (besides of feature freeze)?

File: xbmc/listproviders/MultiProvider.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  virtual void Fetch(std::vector<CGUIListItemPtr> &items) const override;
+  virtual bool IsUpdating() const override;
+  virtual void Reset(bool immediately = false) override;
+  virtual bool OnClick(const CGUIListItemPtr &item) override;
+  virtual bool OnInfo(const CGUIListItemPtr &item) override;
+  virtual bool OnContextMenu(const CGUIListItemPtr &item) override;
+  
+protected:
+  std::vector<IListProviderPtr> m_providers;
+  mutable std::map<CGUIListItem*, IListProviderPtr> m_itemMap;
```
Comment (b-pass) [inline xbmc/listproviders/MultiProvider.h]: The idea is that the MultiProvider needs to know which sub-provider will be sent Click/Info/Context messages so it needs to keep track of which content element belong to which sub-provider.

The only way I see to maintain that mapping is to build it during Fetch().  Unfortunately, Fetch is const in the base class.

I could maybe change that without too much impact on other code.  Would that be preferable to a mutable here?

Comment (b-pass) [inline xbmc/listproviders/MultiProvider.h]: I went ahead and made that change an squashed it in

**Comment by b-pass** [comment]
Fixed conflicts and squashed

**Comment by b-pass** [comment]
Is there anything else I need to get this merged?  I don't want it to get forgotten and miss the boat for v19....

**Comment by ronie** [comment]
jenkins build this please

File: xbmc/listproviders/MultiProvider.cpp [hunk 3]
```diff
@@ -0,0 +1,109 @@
+/*
+ *      Copyright (C) 2013 Team XBMC
+ *      http://xbmc.org
```
Comment (xhaggi) [inline xbmc/listproviders/MultiProvider.cpp]: Please update the copyright in all files you've touched.

```
*      Copyright (C) 201X-2017 Team Kodi
*      http://kodi.tv
```

**Review by xhaggi — CHANGES_REQUESTED** [review]

File: xbmc/listproviders/MultiProvider.cpp [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  CSingleLock lock(m_section);
+  std::vector<CGUIListItemPtr> subItems;
+  items.clear();
+  m_itemMap.clear();
+  for (auto const& provider : m_providers)
+  {
+    provider->Fetch(subItems);
+    for (auto& item : subItems)
+    {
+      m_itemMap[item.get()] = provider;
```
Comment (tamland) [inline xbmc/listproviders/MultiProvider.cpp]: This doesn't look very safe. There is no guarantee other providers aren't going to kill these items at any time, so you should hold a shared pointer.

Comment (b-pass) [inline xbmc/listproviders/MultiProvider.cpp]: I never access the items from the pointers held in this map.  When events come in they have a shared pointer, the map just directs which provider should receive the call.

File: xbmc/listproviders/MultiProvider.cpp (hunk 3 above)
Comment (b-pass) [inline xbmc/listproviders/MultiProvider.cpp]: @xhaggi Done.

File: xbmc/listproviders/MultiProvider.cpp (hunk 4 above)
Comment (b-pass) [inline xbmc/listproviders/MultiProvider.cpp]: @tamland I changed the map definition to use a void* for this key, which will help discourage someone from accidentally using the key as a pointer.

Comment (tamland) [inline xbmc/listproviders/MultiProvider.cpp]: Well, it's still storing possibly dangling pointer which is being accessed by `.find`. Are you sure this is legal and safe?

Comment (Paxxi) [inline xbmc/listproviders/MultiProvider.cpp]: I'm with @tamland on this one. It's a neat trick but it's going to blow up sooner or later, it's also quite non-intuitive to use a pointer address as the key value. There should be a better way to keep track of the items.

Comment (b-pass) [inline xbmc/listproviders/MultiProvider.cpp]: @tamland @Paxxi Alright, I've changed it to use a size_t for keys.

**Comment by tamland** [comment]
jenkins build this please

**Pull request merged**